src/main/scala/abalone/variant/Variant.scala

```diff
@@ -198,68 +198,80 @@ abstract class Variant private[variant] (
+            C: lineDir: upRight, sideDir: right
+            D: lineDir: right, sideDir: upRight
+
         - else:
           - push :
             - dest contains a marble
-              - apply the direction from dest until there is an empty square or being off the board
-                - in case of being off the board, increment the score (capture = true)
-              - do the line move (orig to dest)
+              - move dest marble to the next available square (or off the board)
+              - do the line move
           - line move :
-            - move from orig to dest, without considering the direction
+            - move from orig to dest
   */
   private def piecesAfterAction(pieces: PieceMap, orig: Pos, dest: Pos): (PieceMap) = {
-    val origToDestDir: Direction = Pos.dirFromString(orig.dir(dest))
-
+    val origToDestDirString = orig.directionString(dest)
+    val origToDestDir: Direction = Pos.dirFromString(origToDestDirString)
+ 
     if (isSideMove(orig, dest)) {
-      val lineDir: Option[Direction] = Pos.possibleLineDirsFromSideMoveDir(origToDestDir).flatMap {
-        direction => direction(orig) match {
-          case (Some(squareInLine)) if(pieces.contains(squareInLine) && pieces(squareInLine).player == pieces(orig).player) => Some(direction)
-          case _ => None
-        }
-      }.headOption
-      val sideDir: Option[Direction] = Pos.possibleLineDirsFromSideMoveDir(origToDestDir).flatMap {
-        direction => direction(orig) match {
-          case (Some(squareInLine)) if(!pieces.contains(squareInLine)) => Some(direction)
-          case _ => None
+      val potentialLineDirs = Pos.potentialLineDirsFromSideMoveDir(origToDestDir)
+
+      potentialLineDirs.foreach { lineDir =>
+        val lineDir2ndPos: Option[Pos] = lineDir(orig)
+        val sideDirs = Pos.deducePotentialSideDirs(origToDestDir, lineDir)
+
+        lineDir2ndPos.foreach { secondPos =>
+          sideDirs.foreach { sideDir =>
+            sideDir(secondPos).foreach { side2ndPos =>
+              if (pieces.contains(secondPos))
+                if (side2ndPos == dest && sideDir(orig).isDefined)
+                  return pieces +
+                    (sideDir(orig).get -> pieces(orig)) - orig +
+                    (dest -> pieces(secondPos)) - secondPos
+                else {
+                  val lineDir3rdPos: Option[Pos] = lineDir(secondPos)
+                  lineDir3rdPos.foreach { thirdPos =>
+                    sideDir(thirdPos).foreach { side3rdPos =>
+                      if (pieces.contains(thirdPos) && side3rdPos == dest && sideDir(orig).isDefined) {
+                        return pieces +
+                          (sideDir(orig).get -> pieces(orig)) - orig +
+                          (side2ndPos -> pieces(secondPos)) - secondPos +
+                          (dest -> pieces(thirdPos)) - thirdPos
+                      }
+                    }
+                  }
+                }
+            }
+          }
```

Comment:
Yeah I haven't used `foreach` much myself in scala I don't think its a functional conception.

I asked copilot about refactoring this chunk of code and this is what it said:
> To refactor the provided code to reduce the use of nested foreach loops, you can use more functional constructs such as flatMap and filter. Here's an example of how you can refactor it:

```
val potentialLineDirs = Pos.potentialLineDirsFromSideMoveDir(origToDestDir)

potentialLineDirs.flatMap { lineDir =>
  val lineDir2ndPos = lineDir(orig)
  val sideDirs = Pos.deducePotentialSideDirs(origToDestDir, lineDir)

  lineDir2ndPos.toList.flatMap { secondPos =>
    sideDirs.flatMap { sideDir =>
      val side2ndPos = sideDir(secondPos)

      side2ndPos.collect {
        case destPos if pieces.contains(secondPos) && sideDir(orig).isDefined =>
          if (destPos == dest) {
            return pieces +
              (sideDir(orig).get -> pieces(orig)) - orig +
              (dest -> pieces(secondPos)) - secondPos
          } else {
            lineDir(secondPos).toList.flatMap { thirdPos =>
              sideDir(thirdPos).collect {
                case destPos3 if pieces.contains(thirdPos) && destPos3 == dest && sideDir(orig).isDefined =>
                  return pieces +
                    (sideDir(orig).get -> pieces(orig)) - orig +
                    (destPos -> pieces(secondPos)) - secondPos +
                    (dest -> pieces(thirdPos)) - thirdPos
              }
            }
          }
      }
    }
  }
}
```

> In this refactoring:
> - flatMap is used to handle the optional values and iterate over collections in a nested manner.
> - collect is used to filter and transform elements that meet specific conditions.
> - The return statements inside the blocks remain for early exits when conditions are met.

> This refactoring makes the code more concise and removes deeply nested structures, enhancing readability.